### ISSUE TITLE
NTFS accessor not working correctly

### ISSUE THREAD

**Issue by angry-bender** [Velocidex/velociraptor#2110]
**Describe the bug**
Numerous NTFS accessor and MFT bugs have been noted in VR 0.6.6-1. This includes the following

1. When parsing items that show in MFTEcmd.exe as `.\PathUnknown\Director with ID <string>` velociraptor shows these files under the root directory of a given drive
2. When using the NTFS accessor recent folders are not showing, meaning the auto selector needs to be used instead

**To Reproduce**
Steps to reproduce the behavior (For Issue 1):
1. Run a MFT hunt for a file name
2. Filenames show as C:\<Filename>
3. Verify with MFTECmd.exe and the same files show as .\PathUnknown\Directory with ID <random 0x>

Steps to reproduce the behavior (For Issue 2):
1. Search for a folder name that has been recently created (We've found between 1-3 months) using the NTFS file browser
2. The directory will not show
3. Compare the same directories with the auto accessor

**Expected behavior**
Issue 1 - An output which either shows the directory, or pathunknwon
Issue 2 - NTFS parser finds the directories

**Desktop (please complete the following information):**
 - Client OS: Microsoft Windows Server 2016 Standard10.0.14393 Build 14393
 - VR Server
 "name":"velociraptor"
  "version":"0.6.6-1"
  "compiler":"go1.18.4"

**Additional context**
Nil

**Comment by scudette** [comment]
Just to double check - I created a directory then browsed into it with the ntfs accessor in the VFS maybe < 30 sec later and could see the new directory
![image](https://user-images.githubusercontent.com/3856546/192214885-0ef87bc8-58da-4598-8ec7-18d38afe0153.png)

**Comment by angry-bender** [comment]
Thanks @scudette ,

For Issue 2
In this case, we can see several directories available under VFS, but NFTS isn't parsing the same directories for some reason. In this environment, it seems to be from files created on a secondary drive in the last 3 months.

Looking at the MFTECMD output, it looks like this is closer to Issue 1, The directory is still "in use" or allocated, but even with MFTEcmd its reporting it as `\PathUnknown\Directory with ID <random  0x0>\directory name`

**Comment by svch0stz** [comment]
Perhaps for the first issue instead of returning something like `C:\random_dir\example.txt ` when its failed MFT Parent ID resolution, it could be something like `C:\UNKNOWN-DIRECTORY\random_dir\example.txt `
or just copy Eric's format of `.\PathUnknown\Directory with ID <ID>\random_dir\example.txt `

**Comment by scudette** [comment]
What might explain it is that when you use the VFS (and the ntfs accessor), Velociraptor goes **down** the directory hierarchy by using the $I30 index (basically the same way the filesystem does it). On the other hand when we reconstruct the path in the MFT partifact we go **up** the directory hierarchy by following the parent.

Now normally this should be the same - the parent MFT ID for every MFT entry should point to the containing directory and that directory should in turn contain all the child entries in its $I30 index. 

It looks like something weird is happening in your case - the directory might not actually contain the file, but the file might point back at it - there is some inconsistency there . I am just surprised that such an inconsistency exists with an allocated file? Are you sure the file is marked as allocated?

**Comment by angry-bender** [comment]
Its definitely unusual. 

I can confirm in this case the file is allocated, Just verified live in PowerShell as well. Is there something I can look for in the $I30 index which could help?

**Comment by scudette** [comment]
You can use this artifact https://docs.velociraptor.app/artifact_references/pages/windows.ntfs.i30/ to easily show the $i30 from all relevant directories above the suspected file.

You can also use https://docs.velociraptor.app/artifact_references/pages/windows.ntfs.mft/ with the AllNtfs set to TRUE to see detailed information about the MFT entry of the file in question.

**Comment by angry-bender** [comment]
> Perhaps for the first issue instead of returning something like `C:\random_dir\example.txt ` when its failed MFT Parent ID resolution, it could be something like `C:\UNKNOWN-DIRECTORY\random_dir\example.txt ` or just copy Eric's format of `.\PathUnknown\Directory with ID <ID>\random_dir\example.txt `

I think this is a good idea, as it could minimise the chances of misinterpretation there. I know some people could interpret the output to mean significant findings like filestaging or credential theft showing on the root directory of the drive.



> You can use this artifact https://docs.velociraptor.app/artifact_references/pages/windows.ntfs.i30/ to easily show the $i30 from all relevant directories above the suspected file.
> 
> You can also use https://docs.velociraptor.app/artifact_references/pages/windows.ntfs.mft/ with the AllNtfs set to TRUE to see detailed information about the MFT entry of the file in question.

I'll try and pull that again, the last one timed out there

**Comment by scudette** [comment]
Just to add more information: It looks like some files exist with **different** parent MFT entry in multiple $FILE_NAME streams. This means the same file belongs in multiple directories. Velociraptor's path reconstruction only picks the first one but I guess any of the other paths are equally correct. These files are allocated but happen to exist in two directories at the same time legitimately.

Here is the output from TSK illustrating this:
```
 .\ffind.exe  \\.\c: 3440
/Program Files/WindowsApps/Microsoft.VP9VideoExtensions_1.0.51171.0_x64__8wekyb3d8bbwe/Assets/contrast-black/AppList.targetsize-80_contrast-black.png
/Program Files/WindowsApps/Microsoft.WebpImageExtension_1.0.52351.0_x64__8wekyb3d8bbwe/Assets/AppList.targetsize-80.png

 .\istat.exe  \\.\c: 3440
MFT Entry Header Values:
Entry: 3440        Sequence: 50
Base File Record: 23167
$LogFile Sequence Number: 6141651614
Allocated File
Links: 0

$FILE_NAME Attribute Values:
Flags: Archive
Name: AppList.targetsize-80_contrast-black.png
Parent MFT Entry: 135256        Sequence: 1
Allocated Size: 4096    Actual Size: 438
Created:        2021-06-05 07:31:49.068332000 (Pacific Daylight Time)
File Modified:  2021-06-05 07:31:49.068332000 (Pacific Daylight Time)
MFT Modified:   2022-07-30 14:48:02.422799900 (Pacific Daylight Time)
Accessed:       2022-07-30 21:41:52.275979500 (Pacific Daylight Time)

$FILE_NAME Attribute Values:
Flags: Archive
Name: AppList.targetsize-80.png
Parent MFT Entry: 352619        Sequence: 5
Allocated Size: 4096    Actual Size: 438
Created:        2021-06-05 07:31:49.068332000 (Pacific Daylight Time)
File Modified:  2021-06-05 07:31:49.068332000 (Pacific Daylight Time)
MFT Modified:   2022-09-13 21:44:58.809784100 (Pacific Daylight Time)
Accessed:       2022-09-13 08:03:23.448366100 (Pacific Daylight Time)

Attributes:
Type: $FILE_NAME (48-0)   Name: N/A   Resident   size: 146
Type: $FILE_NAME (48-1)   Name: N/A   Resident   size: 116

```

**Comment by scudette** [comment]
More information: The ntfs path reconstruction code has built in safety against corrupted filesystems which result in cycles. It follows only 10 parents and then gives up:

https://github.com/Velocidex/go-ntfs/blob/3a8f0f1aeae9fb6367987f1e5343c6de42a9ffd0/parser/utils.go#L42

so if a directory is deeper than 10 levels in - the code will give up and put it at the root.

**Comment by scudette** [comment]
I just submitted #2125 that adds the error message in a virtual directory so it is clear that the path reconstruction failed. It also adds a check for sequence numbers when traversing parents in the path. I did some preliminary testing and for allocated files it is always the case that the sequence numbers match so this change should not break allocated file reconstruction - it does however eliminate a large amount of unallocated file reconstruction which I guess is good because these paths are probably wrong.

Please test with the latest CI binaries to see if the results are a bit easier to understand now.

**Comment by scudette** [comment]
I just submitted #2134 which adds a new field to parse_ntfs() called Links. This is an array of full paths. Since NTFS supports hard links it is possible and quite common for a file to exist in multiple directories with multiple filenames

This can easily be observed with notepad.exe for example:
```
C:> fsutil.exe hardlink list "C:/Windows/notepad.exe"
\Windows\WinSxS\amd64_microsoft-windows-notepad_31bf3856ad364e35_10.0.22000.653_none_6a6258f1f542ccd6\notepad.exe
\Windows\System32\notepad.exe
\Windows\notepad.exe
```

It is also possible to create a hardlink for any file using fsutil. The new column contains the list of full paths that belong to the same file so we can filter them all at once.

**Comment by scudette** [comment]
Should be fixed in head #2134
